### PR TITLE
Perform Transitive Reduction on Graph

### PR DESCRIPTION
**Pull request by mitchellh**
This adds a new transformer to perform a [transitive reduction](http://en.wikipedia.org/wiki/Transitive_reduction) on the graph. We make some pretty edge-y graphs, and this should help simplify them considerably for debugging. The changes made in this PR:

* Add `*AcyclicGraph.TransitiveReduction()` to perform the actual transitive reduction

* Add `TransitiveReductionTransformer` that does the transitive reduction transform

### CODE REVIEW

**Comment by armon** [comment]
LGTM. This is so much more *infinitely* sane that our previous approach...

File: terraform/transform_transitive_reduction.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package terraform
+
+// TransitiveReductionTransformer is a GraphTransformer that performs
+// finds the transitive reduction of the graph. For a definition of
+// transitive reduction, see Wikipedia.
```
Comment (phinze) [inline terraform/transform_transitive_reduction.go]: Hahahah - did you just http://www.letmewikipediathatforyou.com/?q=Transitive%20Reduction in a code comment?

Comment (mitchellh) [inline terraform/transform_transitive_reduction.go]: Yes, I figured if you don't know what a transitive reduction is, then saying "maximally reduces the edges in a graph to ensure the same reachability" probably still wouldn't make sense. ;)

**Comment by phinze** [comment]
Down the road, we may want to consider using this as a base for a mechanism for providing feedback to the user on redundant  `depends_on` in their config. Falls pretty squarely in the "nice to have" column though.

**Comment by mitchellh** [comment]
@phinze That is an interesting idea. Future `terraform lint` perhaps? ;)

**Comment by mitchellh** [comment]
Real world improvement from an anonymous not-small user:

Before: 181 vertexes, 541 edges
After: 181 vertexes, 350 edges

Solid.

**Pull request merged**